nasl/nasl_packet_forgery.c

```diff
@@ -570,9 +570,9 @@ get_tcp_element (lex_ctxt *lexic)
   else if (!strcmp (element, "data"))
     {
       retc = alloc_typed_cell (CONST_DATA);
-      retc->size = UNFIX (ip->ip_len) - ntohl (tcp->th_off) * 4;
+      retc->size = UNFIX (ip->ip_len) - tcp->th_off * 4;
       retc->x.str_val = g_malloc0 (retc->size);
-      bcopy (tcp + ntohl (tcp->th_off) * 4, retc->x.str_val, retc->size);
+      bcopy (tcp + tcp->th_off * 4, retc->x.str_val, retc->size);
```

Comment:
In nasl/nasl_packet_forgery.c the following is used:

```
bcopy ((char *) tcp + tcp->th_off * 4, retc->x.str_val, retc->size);
```

I can't tell about the difference between with / without `(char *)` but might worth a look.

Besides the above i start doing some testing now.

Reply:
You are right. Nice catch. It has to be `(char *)tcp`. Else it would be incremented in steps of 20 bytes (tcphdr size).

Reply:
Could this also be responsible that "crap" data is included in the response? I'm asking because on a `get_tcp_element` to `data` i'm getting something like e.g. the following back on a TCP packet without any data field in the TCP packet:

```
lib  nasl-Message: 10:22:10.396: 0x00:  00 00 00 00 71 00 00 00 00 00 00 00 20 56 F8 7B    ....q....... V.{
0x10:  CD 55 00 00                                        .U.. 
```

Reply:
Confirmed that 3e1a3db295f0b649bdef22868fdb148513e9f2af solves the data extraction.